### PR TITLE
feat(mobile): add initialPage prop to Carousel

### PR DESCRIPTION
**Pull request by caitlin-coyiuto-cb**
[CB1-2048](https://linear.app/coinbase/issue/CB1-2048)

## What changed? Why?

- Add an uncontrolled `initialPage` prop to the cds-mobile `Carousel` so consumers can open it on a non-zero page with no animation.
- Seed `activePageIndex` at init + a one-time, `didInitRef`-guarded non-animated seek on first measurement (instant `.set()`, no slide from page 0); reads `initialPage` via a ref since data can arrive async.
- No `onChangePage` on mount; clamps out-of-range; routes through `findNearestLoopOffset` when looping; a later `goToPage`/drag still wins.
- Replaces the consumer-side `requestAnimationFrame` + `goToPage` workaround (clears `TODO(CB1-2026)`); unblocks the RN adoption PR (PR 2).

### Root cause (required for bugfixes)

N/A — feature, not a bugfix.

## UI changes

Adds an "Initial Page (opens on page 3)" example to the mobile Carousel stories. Web unaffected.

https://github.com/user-attachments/assets/41ba9a4b-50d4-4305-9f22-bddb79584778

https://github.com/user-attachments/assets/217eaaec-5a6a-49a0-9d7d-56eadb764968

### How has it been tested?

- [x] Unit tests
- [ ] Interaction tests
- [ ] Pseudo State tests
- [ ] Manual - Web
- [ ] Manual - Android (Emulator / Device)
- [X] Manual - iOS (Emulator / Device)

### Testing instructions

**Automated:** New `Initial Page` tests in `Carousel.test.tsx` — opens on `initialPage`; `onChangePage` not called on mount; out-of-range clamps (`99→6`, `-5→0`); later `goToPage` and drag override the initial page. Full file 97/97 pass; `mobile` typecheck + lint clean.

**Manual:** `yarn nx run expo-app:ios` → open the Carousel route → the "Initial Page (opens on page 3)" example lands on page 3 with no slide from page 1; dragging/pagination still work.

## Illustrations/Icons Checklist

N/A — no changes under `packages/illustrations/**` or `packages/icons/**`.

## Change management

type=routine
risk=low
impact=sev5

automerge=false

### CODE REVIEW

**Comment by cb-heimdall** [comment]
<h1>✅ Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
✅
<code>1/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td><code>2</code> if repo is sensitive</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>1</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
<tr>
<td><code>CODEOWNERS</code></td>
<td>
✅ See below
</td>
<td></td>
</tr>
</table>

<h2>
✅
<code>CODEOWNERS</code>
</h2>
<table>
<tr>
<th>Code Owner</th>
<th>Status</th>
<th>Calculation</th>
</tr>
<tr>

<td>ui-systems-eng-team</td>

<td>
✅
<code>1/1</code>
</td>


<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
Additional CODEOWNERS Requirement
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Sum</strong></td>
<td>
<strong>
0
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>From <code>CODEOWNERS</code></td>
<td>1</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

File: packages/web/CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,10 @@ All notable changes to this project will be documented in this file.
 
 <!-- template-start -->
 
+## 9.23.0 ((8/27/2026, 05:05 PM PST))
+
+This is an artificial version bump with no new change.
+
```
Comment (hcopp) [inline packages/web/CHANGELOG.md:16]: @caitlin-coyiuto-cb are we able to get web parity here? :pray:

File: packages/mobile/src/carousel/Carousel.tsx [hunk 2]
```diff
@@ -188,6 +188,15 @@ export type CarouselBaseProps = SharedProps &
...
      */
     snapMode?: 'item' | 'page';
+    /**
+     * The page index to open the carousel on initially. This is uncontrolled: it seeds the
+     * starting page once (on first measurement) and the carousel manages its own page after
+     * that — passing a new value later will not move it. The carousel opens directly on this
+     * page with no animation, and `onChangePage` does not fire on mount. Out-of-range values
+     * are clamped to `[0, totalPages - 1]`.
+     * @default 0
+     */
```
Comment (hcopp) [inline packages/mobile/src/carousel/Carousel.tsx:196]: nit: can we shorten this so it is readable on our docs? Maybe something like

```
/**
 * Zero-based page index to show on first layout. Out-of-range values are clamped.
 * @default 0
 */
initialPageIndex?: number;
```

File: packages/mobile/src/carousel/Carousel.tsx [hunk 3]
```diff
@@ -748,6 +764,34 @@ export const Carousel = memo(
...
+      updateVisibleCarouselItems(seekOffset);
+    }, [
+      hasCalculatedDimensions,
+      totalPages,
+      pageOffsets,
+      isLoopingActive,
+      loopLength,
+      animationApi.x,
+      updateVisibleCarouselItems,
+    ]);
```
Comment (hcopp) [inline packages/mobile/src/carousel/Carousel.tsx]: What if we had a single useLayoutEffect instead, maybe something like

```
    useLayoutEffect(() => {
      if (didInitRef.current || totalPages === 0) return;

      didInitRef.current = true;
      applyPage(initialPageIndex); // here we could modify applyPage to have some sort of toggle for notifying and animation
    }, [totalPages, initialPageIndex, applyPage]);
```

File: packages/mobile/src/carousel/Carousel.tsx [hunk 4]
```diff
@@ -582,12 +592,18 @@ export const Carousel = memo(
     } = mergedProps;
     const carouselScrollX = useRef(0);
 
+    // Kept in sync each render so the one-time seek effect reads the value current at first
+    // measurement (the target page may arrive asynchronously after mount).
+    const initialPageRef = useRef(initialPage);
+    initialPageRef.current = initialPage;
```
Comment (hcopp) [inline packages/mobile/src/carousel/Carousel.tsx]: How big of a concern is it if the prop is changed between initial load and when values are set? I think it should be fine without a ref here.

**Review by hcopp — COMMENTED** [review]
Thanks for the PR! Had some feedback

File: packages/mobile/src/carousel/Carousel.tsx (hunk 4 above)
Comment (cb-ekuersch) [inline packages/mobile/src/carousel/Carousel.tsx]: this will update the ref when the initial page changes which shouldn't matter unless the component remounts and in that instance the useRef will evaluate again

File: packages/mobile/src/carousel/Carousel.tsx [hunk 5]
```diff
@@ -582,12 +592,18 @@ export const Carousel = memo(
...
+    const didInitRef = useRef(false);
+
     const animationApi = useSpring({
       x: carouselScrollX.current,
       config: animationConfig,
     });
 
-    const [activePageIndex, setActivePageIndex] = useState(0);
+    const [activePageIndex, setActivePageIndex] = useState(() => Math.max(0, initialPage ?? 0));
```
Comment (cb-ekuersch) [inline packages/mobile/src/carousel/Carousel.tsx:599]: we should probably use the value from the ref here in case the initialPage prop does in fact change. If that were the case then the activePage state would also change

File: packages/mobile/src/carousel/Carousel.tsx [hunk 6]
```diff
@@ -748,6 +764,34 @@ export const Carousel = memo(
...
     ]);
 
+    // One-time, non-animated seek to `initialPage` on the first successful measurement.
+    // activePageIndex is already seeded from `initialPage`, so this only aligns the scroll
+    // position and visible items — no onChangePage fires.
+    useEffect(() => {
+      if (didInitRef.current || !hasCalculatedDimensions || totalPages === 0) return;
+      didInitRef.current = true;
+
+      const target = Math.max(0, Math.min(totalPages - 1, initialPageRef.current ?? 0));
```
Comment (cb-ekuersch) [inline packages/mobile/src/carousel/Carousel.tsx]:
```
const target = Math.max(0, Math.min(totalPages - 1, initialPageRef.current ?? 0));
      if (target === 0) return;

      const baseOffset = pageOffsets[target];
      const seekOffset = isLoopingActive
        ? findNearestLoopOffset(carouselScrollX.current, [baseOffset], loopLength).offset
        : baseOffset;

      carouselScrollX.current = seekOffset;
      animationApi.x.set(seekOffset);
      updateVisibleCarouselItems(seekOffset);
```

This logic looks very similar to what we do on normal page navigation. Is there a way to cleanly share the logic?

File: packages/mobile/src/carousel/Carousel.tsx (hunk 3 above)
Comment (cb-ekuersch) [reply packages/mobile/src/carousel/Carousel.tsx]: +1 the useLayoutEffect sounds like the correct React primitive to use since our side-effect deals with visual measurements

File: packages/web/src/carousel/__stories__/Carousel.stories.tsx [hunk 7]
```diff
@@ -658,12 +658,28 @@ const LoopingExamples = () => (
...
+  </Carousel>
+);
+
 export const All = () => (
   <VStack gap={2}>
     <BasicExamples />
     <CustomComponentsExample />
     <CustomStylesExample />
     <AnimatedPaginationExample />
+    <InitialPageExample />
```
Comment (hcopp) [inline packages/web/src/carousel/__stories__/Carousel.stories.tsx]: Could we move this down to the bottom? that helps our visual regression testing to compare before vs after

File: packages/mobile/src/carousel/Carousel.tsx [hunk 8]
```diff
@@ -792,6 +805,19 @@ export const Carousel = memo(
...
+    // Open on `initialPage` once, on the first layout, with no animation. activePageIndex is
+    // already seeded from `initialPage`, so goToPage realigns scroll/visibility without firing
+    // onChangePage on mount.
+    useLayoutEffect(() => {
+      if (hasAppliedInitialPageRef.current || totalPages === 0) return;
+      hasAppliedInitialPageRef.current = true;
+
+      const targetPage = initialPage ?? 0;
+      if (targetPage > 0) {
+        goToPage(targetPage, { animate: false });
```
Comment (hcopp) [inline packages/mobile/src/carousel/Carousel.tsx:814]: Nice!

File: packages/mobile/src/carousel/Carousel.tsx [hunk 9]
```diff
@@ -792,6 +805,19 @@ export const Carousel = memo(
       ],
     );
 
+    // Open on `initialPage` once, on the first layout, with no animation. activePageIndex is
+    // already seeded from `initialPage`, so goToPage realigns scroll/visibility without firing
+    // onChangePage on mount.
```
Comment (hcopp) [inline packages/mobile/src/carousel/Carousel.tsx]: nit: this comment feels distracting, I think the code below is self sufficient

File: packages/mobile/src/carousel/Carousel.tsx (hunk 5 above)
Comment (hcopp) [reply packages/mobile/src/carousel/Carousel.tsx:599]: @caitlin-coyiuto-cb I think we need to clamp this with the max as well right?

Comment (caitlin-coyiuto-cb) [reply packages/mobile/src/carousel/Carousel.tsx:599]: I don't think we can here on first render since it'd be 0? totalPages gets derived on line 725 later. Also have tests for clamping on last page (if initialPage > last page) and negative initialPage

Comment (hcopp) [reply packages/mobile/src/carousel/Carousel.tsx:599]: Yeah I was wondering if we needed to set activePageIndex when we set hasAppliedInitialPageRef to true. I suppose goToPage does clamp automatically but I wasn't sure if there is an interim side effects if this value was way off.

**Review by hcopp — APPROVED** [review]
LGTM, thanks for all the updates
